shell/common/shell_unittests.cc

```diff
@@ -3721,6 +3721,74 @@ TEST_F(ShellTest, UsesPlatformMessageHandler) {
   DestroyShell(std::move(shell));
 }
 
+TEST_F(ShellTest, SpawnWorksWithOnError) {
```

Comment:
Now that you've reworked this, this is less of a concern for me.  I would still feel better if instead of having a global variable we had a variable associated with an engine.  It's still possible that one person edits `log_unhandled_exception` one one UI thread while another UI thread is reading from it.  That wouldn't happen in the spawn case since the 2 shells are sharing the ui thread.  Maybe you can copy the value of `log_unhandled_exception` at some point to make sure they have their own copy?

Maybe it could just live in the platform configuration?  That sort of makes more sense than having a global variable anyways, no?

Reply:
I'm not sure I'm following.

I'm not making any changes to the way `log_unhandled_exception` works. This isn't public API though. There's no way for an embedder to actually make this mistake, right?

Reply:
hah. There's another thing called `log_message_callback` in `UIDartState` and I had just been looking at that. I see now.

The thing that makes this all work is that the function is actually going to call the current isolate from UIDartState::Current. But we can't fully encode that here because tonic doesn't know about UIDartState.

Reply:
IOW, someone editing UIDartState can mess this up, but no embedder can mess it up.

Reply:
So `SetUnhandledExceptionReporter` is called in `UIDartState::UIDartState` right?  If a user creates a second engine while the first engine was reading the value of `log_unhandled_exception` you could get undefined behavior.  Even though you are setting the value to the exact same value there is no guarantee that at all points in time the value is correct.  For that guarantee you'd have to make the function pointer atomic.

Reply:
Using a std::atomic for this.

Reply:
It might make sense to call `SetUnhandledExceptionReporter` from `DartVMInitializer::Initialize` instead of doing it from the `UIDartState` constructor.

This will ensure that the global exception reporter function is consistently set up for the entirety of the Dart VM's lifetime.   I've been using `DartVMInitializer::Initialize` to do setup for another global callback that sends trace events to the Dart timeline.

Reply:
I'm not really sure how that would work - we could have `DartVMInitializer` be the thing that registers with tonic, but we'd still need UIDartState to register with the intializer right?

Reply:
`DartVMInitializer::Initialize` would call `tonic::SetUnhandledExceptionReporter(ReportUnhandledException)`

`ReportUnhandledException` is now a static function that calls `UIDartState::Current()` and then uses the `UIDartState` API if available.  It is no longer a lambda or method tied to a `UIDartState` instance.

`ReportUnhandledException` could be declared in `ui_dart_state.h` to be publicly accessible to `DartVMInitializer`.  Or its implementation could be moved to `dart_vm_initializer.cc`

I think this would be preferable in order to avoid global side effects in the `UIDartState` constructor.

Reply:
Ahh of course, ok that makes perfect sense.